### ISSUE TITLE
TestInteractiveMarkerServerWithMarkers tests are failing

### ISSUE THREAD

**Issue by piraka9011** [ros-visualization/interactive_markers#56]
From the nightly build farm, [armhf](https://ci.ros2.org/view/nightly/job/nightly_linux-armhf_debug/lastBuild/testReport/junit/interactive_markers/TestInteractiveMarkerServerWithMarkers/) and [x86_64](https://ci.ros2.org/view/nightly/job/nightly_linux_debug/1402/testReport/junit/interactive_markers/TestInteractiveMarkerServerWithMarkers/) builds are failing.

Multiple tests fail with a similar stack trace:

------

**Stacktrace**

unknown file
`C++ exception with description "failed to initialize wait set: the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at /home/jenkins-agent/workspace/nightly_linux-armhf_debug/ws/src/ros2/rcl/rcl/src/rcl/wait.c:130" thrown in SetUp().`

**Standard Output**

unknown file
`C++ exception with description "failed to initialize wait set: the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at /home/jenkins-agent/workspace/nightly_linux-armhf_debug/ws/src/ros2/rcl/rcl/src/rcl/wait.c:130" thrown in SetUp().`

------

Reproduced it locally. 

@wjwwood @dgossow any chance you guys know why that is? I wasn't able to find where this issue could be coming from?

**Comment by wjwwood** [comment]
@piraka9011, I think @dgossow is unlikely to know since you're speaking about the ROS 2 port of interactive markers. @jacobperron ported this code, so he might know though.

**Comment by nuclearsandwich** [comment]
I have been tracking this in https://github.com/ros2/build_cop/issues/254 and have a bit of additional info there. I'm not sure whether changes will need to be made here (if the problem lies with the test implementation) or if there's an issue in rclcpp that these tests are exposing.

**Comment by jacobperron** [comment]
`rclcpp::init` and `rclcpp::shutdown` are called once each:

https://github.com/ros-visualization/interactive_markers/blob/674d15927caf41684833ec1f843cbb03781f148a/test/interactive_markers/test_interactive_marker_server.cpp#L113-L121

But, there are test above that don't use the test fixture and call `rclcpp::init` and `rclcpp::shutdown` at the start and end of each test. E.g.

https://github.com/ros-visualization/interactive_markers/blob/674d15927caf41684833ec1f843cbb03781f148a/test/interactive_markers/test_interactive_marker_server.cpp#L88-L108

We might run into a situation where shutdown doesn't get called if there is an uncaught exception raised during the test. But if that was happening I'd expect we'd see other test failures and  a different error message coming from rcl.

In summary, I'm not convinced the reported failure is due to the test code in interactive_markers.

**Comment by nuclearsandwich** [comment]
What's the difference between {SetUp,TearDown} and {SetUp,TearDown}TestCase? It might help to come up with an independent reproduction.

**Comment by jacobperron** [comment]
My understanding is that {SetUp, TearDown}TestCase occurs once per test fixture (ie. before the first test and after the last test associated with the fixture) and the non-TestCase variant runs before and after each test. [Reference](https://github.com/google/googletest/blob/v1.8.x/googletest/docs/advanced.md#sharing-resources-between-tests-in-the-same-test-case)

**Comment by wjwwood** [comment]
I think someone just has to investigate this issue locally.

I don't know why, but based on this output from CI (https://ci.ros2.org/view/nightly/job/nightly_linux_debug/1402/consoleFull and search `TestInteractiveMarkerServerWithMarkers.erase`):

```
03:05:56 1: Running main() from /home/jenkins-agent/workspace/nightly_linux_debug/ws/install/gtest_vendor/src/gtest_vendor/src/gtest_main.cc
03:05:56 1: [==========] Running 10 tests from 2 test cases.
03:05:56 1: [----------] Global test environment set-up.
03:05:56 1: [----------] 2 tests from TestInteractiveMarkerServer
03:05:56 1: [ RUN      ] TestInteractiveMarkerServer.construction_and_destruction
03:05:56 1: [       OK ] TestInteractiveMarkerServer.construction_and_destruction (44 ms)
03:05:56 1: [ RUN      ] TestInteractiveMarkerServer.insert
03:05:56 1: [       OK ] TestInteractiveMarkerServer.insert (13 ms)
03:05:56 1: [----------] 2 tests from TestInteractiveMarkerServer (57 ms total)
03:05:56 1: 
03:05:56 1: [----------] 8 tests from TestInteractiveMarkerServerWithMarkers
03:05:56 1: [ RUN      ] TestInteractiveMarkerServerWithMarkers.erase
03:05:56 1: unknown file: Failure
03:05:56 1: C++ exception with description "failed to initialize wait set: the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at /home/jenkins-agent/workspace/nightly_linux_debug/ws/src/ros2/rcl/rcl/src/rcl/wait.c:130" thrown in SetUp().
03:05:56 1: [  FAILED  ] TestInteractiveMarkerServerWithMarkers.erase (25 ms)
03:05:56 1: [ RUN      ] TestInteractiveMarkerServerWithMarkers.clear
03:05:56 1: [       OK ] TestInteractiveMarkerServerWithMarkers.clear (27 ms)
03:05:56 1: [ RUN      ] TestInteractiveMarkerServerWithMarkers.get_marker_by_name
03:05:56 1: [       OK ] TestInteractiveMarkerServerWithMarkers.get_marker_by_name (27 ms)
03:05:56 1: [ RUN      ] TestInteractiveMarkerServerWithMarkers.set_pose
03:05:56 1: unknown file: Failure
03:05:56 1: C++ exception with description "failed to initialize wait set: the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at /home/jenkins-agent/workspace/nightly_linux_debug/ws/src/ros2/rcl/rcl/src/rcl/wait.c:130" thrown in SetUp().
03:05:56 1: [  FAILED  ] TestInteractiveMarkerServerWithMarkers.set_pose (26 ms)
03:05:56 1: [ RUN      ] TestInteractiveMarkerServerWithMarkers.set_callback
03:05:56 1: unknown file: Failure
03:05:56 1: C++ exception with description "failed to initialize wait set: the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at /home/jenkins-agent/workspace/nightly_linux_debug/ws/src/ros2/rcl/rcl/src/rcl/wait.c:130" thrown in SetUp().
03:05:56 1: [  FAILED  ] TestInteractiveMarkerServerWithMarkers.set_callback (26 ms)
03:05:56 1: [ RUN      ] TestInteractiveMarkerServerWithMarkers.feedback_communication
03:05:56 1: [       OK ] TestInteractiveMarkerServerWithMarkers.feedback_communication (199 ms)
03:05:56 1: [ RUN      ] TestInteractiveMarkerServerWithMarkers.update_communication
03:05:56 1: [INFO] [1575975956.676854376] [mock_interactive_marker_client]: Update received
03:05:56 1: [INFO] [1575975956.697372502] [mock_interactive_marker_client]: Update received
03:05:56 1: [INFO] [1575975956.717709755] [mock_interactive_marker_client]: Update received
03:05:56 1: [       OK ] TestInteractiveMarkerServerWithMarkers.update_communication (88 ms)
03:05:56 1: [ RUN      ] TestInteractiveMarkerServerWithMarkers.get_interactive_markers_communication
03:05:56 1: unknown file: Failure
03:05:56 1: C++ exception with description "failed to initialize wait set: the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at /home/jenkins-agent/workspace/nightly_linux_debug/ws/src/ros2/rcl/rcl/src/rcl/wait.c:130" thrown in SetUp().
03:05:56 1: [  FAILED  ] TestInteractiveMarkerServerWithMarkers.get_interactive_markers_communication (24 ms)
03:05:56 1: [----------] 8 tests from TestInteractiveMarkerServerWithMarkers (442 ms total)
03:05:56 1: 
03:05:56 1: [----------] Global test environment tear-down
03:05:56 1: [==========] 10 tests from 2 test cases ran. (499 ms total)
03:05:56 1: [  PASSED  ] 6 tests.
03:05:56 1: [  FAILED  ] 4 tests, listed below:
03:05:56 1: [  FAILED  ] TestInteractiveMarkerServerWithMarkers.erase
03:05:56 1: [  FAILED  ] TestInteractiveMarkerServerWithMarkers.set_pose
03:05:56 1: [  FAILED  ] TestInteractiveMarkerServerWithMarkers.set_callback
03:05:56 1: [  FAILED  ] TestInteractiveMarkerServerWithMarkers.get_interactive_markers_communication
03:05:56 1: 
03:05:56 1:  4 FAILED TESTS
```

when entering the `TestInteractiveMarkerServerWithMarkers::SetupUp()` for the first test run (`erase`), init has not been called (or shutdown has been called). I would not expect this because `SetUpTestCase()` should have been called before where `init()` is done and no other tests have run before it that could have done `shutdow()`. The only thing to note is the non-fixture tests `construction_and_destruction` and `insert` run before and do their own init/shutdown cycle. But they pass and don't return early, so it should be fine.

Also something strange is that several other instances of this exception occur in `SetUp` but not all of them... It's almost like there's asynchronous init/shutdown happening...

**Comment by jacobperron** [comment]
I believe this was fixed by https://github.com/ros2/rclcpp/pull/947
